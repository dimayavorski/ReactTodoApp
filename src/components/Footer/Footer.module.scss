.actionsList {
  margin: 0;
  padding: 20px 15px 10px 15px;
  display: flex;
  justify-content: space-between;
  list-style-type: none;
  .filter {
    display: flex;
    gap: 15px;
  }
}