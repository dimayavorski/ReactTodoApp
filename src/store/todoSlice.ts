import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { ITodoItem } from '../models/ITodoItem';
import { FilterType } from '../models/FilterType';

const todos: ITodoItem[] = [];
const filter: number = FilterType.ALL;

const todoSlice = createSlice({
	name: 'todos',
	initialState: {
		todos,
		filter,
	},
	reducers: {
		addToDo(state, action: PayloadAction<ITodoItem>) {
			state.todos.push(action.payload);
		},
		removeToDo(state, action: PayloadAction<ITodoItem>) {
			state.todos = state.todos.filter((todo) => todo.id !== action.payload.id);
		},
		toggleToDoComplete(state, action: PayloadAction<ITodoItem>) {
			const toggleTodo = state.todos.find(
				(todo) => todo.id === action.payload.id
			);
			if (toggleTodo) toggleTodo.checked = !toggleTodo.checked;
		},
		updateToDos(state, action: PayloadAction<ITodoItem[]>) {
			state.todos = action.payload;
		},
		updateFilter(state, action: PayloadAction<FilterType>) {
			state.filter = action.payload;
		},
	},
});

export const {
	updateToDos,
	addToDo,
	removeToDo,
	toggleToDoComplete,
	updateFilter,
} = todoSlice.actions;

export default todoSlice.reducer;
