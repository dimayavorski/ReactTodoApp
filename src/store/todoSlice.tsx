import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { ITodoItem } from '../models/ITodoItem';

const todos: ITodoItem[] = [];

const todoSlice = createSlice({
	name: 'todos',
	initialState: {
		todos,
	},
	reducers: {
		addToDo(state, action: PayloadAction<ITodoItem>) {
			state.todos.push(action.payload);
		},
		removeToDo(state, action: PayloadAction<ITodoItem>) {
			state.todos = state.todos.filter((todo) => todo.id !== action.payload.id);
		},
		toggleToDoComplete(state, action: PayloadAction<ITodoItem>) {
			const toggleTodo = state.todos.find(
				(todo) => todo.id === action.payload.id
			);
			if (toggleTodo) toggleTodo.checked = !toggleTodo.checked;
		},
		updateToDos(state, action: PayloadAction<ITodoItem[]>) {
			state.todos = action.payload;
		},
	},
});

export const { updateToDos, addToDo, removeToDo, toggleToDoComplete } =
	todoSlice.actions;
//export const selectCount = (state: RootState) => state.counter.value
export default todoSlice.reducer;
