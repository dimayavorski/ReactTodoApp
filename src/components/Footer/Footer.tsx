import { useAppSelector } from '../../hooks/useAppSelector';
import { useTodos } from '../../hooks/useTodos';
import { ITodoItem } from '../../models/ITodoItem';
import { FooterButton } from '../FooterButton/FooterButton';
import styles from './Footer.module.scss';
import { useAppDispatch } from '../../hooks/useAppDispatch';
import { updateFilter } from '../../store/todoSlice';
import { FilterType } from '../../models/FilterType';
import { useEffect, useState } from 'react';

export function Footer() {
	const { clearCompleted, setFiltered } = useTodos();

	const [active, setActive] = useState(0);
	const todos: ITodoItem[] = useAppSelector((state) => state.todos.todos);
	const dispatch = useAppDispatch();
	const setIncompleteFilter = () => {
		dispatch(updateFilter(FilterType.INCOMPLETE));
		setActive(FilterType.INCOMPLETE);
	};
	const setCompletedFilter = () => {
		dispatch(updateFilter(FilterType.COMPLETED));
		setActive(FilterType.COMPLETED);
	};
	const setAllFilter = () => {
		dispatch(updateFilter(FilterType.ALL));
		setActive(FilterType.ALL);
	};

	useEffect(() => {
		setFiltered();
	}, [active]);

	return (
		<ul className={styles.actionsList}>
			<li>{todos.filter((todo) => !todo.checked).length} items left</li>
			<li className={styles.filter}>
				<FooterButton
					isActive={active == FilterType.ALL}
					clickAction={setAllFilter}
					text="All"
					dataTestId="getAll"
				/>

				<FooterButton
					isActive={active == FilterType.INCOMPLETE}
					clickAction={setIncompleteFilter}
					text="Active"
					dataTestId="getActive"
				/>

				<FooterButton
					isActive={active == FilterType.COMPLETED}
					clickAction={setCompletedFilter}
					text="Completed"
					dataTestId="getCompleted"
				/>
			</li>
			<li>
				<FooterButton
					clickAction={clearCompleted}
					text="Clear Completed"
					dataTestId="clearCompleted"
				/>
			</li>
		</ul>
	);
}
