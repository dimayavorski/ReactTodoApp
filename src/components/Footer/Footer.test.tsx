import { RenderResult } from '@testing-library/react';
import { Footer } from './Footer';
import { Provider } from 'react-redux';
import { store } from '../../store/store';

describe('Footer tests', () => {
	let component: RenderResult;

	beforeEach(() => {
		<Provider store={store}>
			component = render(
			<Footer />
			);
		</Provider>;
	});

	it('should match snapshot', () => {
		expect(component).toMatchSnapshot();
	});
});
