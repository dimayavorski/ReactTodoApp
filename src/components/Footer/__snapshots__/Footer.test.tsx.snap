// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer tests should match snapshot 1`] = `undefined`;
